### DISCUSSION TITLE
Minimal Example shows `exec: /hello: no such file or directory`

### DISCUSSION THREAD

**Discussion by p3achyjr** [moby/moby#45226] [Q&A] [answered]
### Description

I am trying to set up Docker, but am having trouble with basic examples. I created a folder with the following structure:

```
hello/
    main.cc
    a.out
    Dockerfile
```

And my Dockerfile:

```
FROM amd64/alpine:3.14
COPY a.out /
CMD ["/a.out"]
```
I run `sudo docker build . -t hello` and it works, however, every time I run `sudo docker run -it --user 8080:80 hello:latest`, I get the error

```
exec: /a.out: no such file or directory
```

To make things weirder, I cloned https://github.com/docker-library/hello-world.git, extracted the files under `amd64`, and built an image with just those, and it worked. The permissions look the same. Really confused here :/

I am running on a GCP VM with Debian.

### Reproduce

1. Create folder with structure 

```
hello/
    main.cc
    a.out
    Dockerfile
```

main.cc:
```
#include <stdio.h>
int main(int argc, char** argv) {
  printf("Hello!\n");
  return 0;
}
```

Dockerfile
```
FROM amd64/alpine:3.14
COPY hello /
CMD ["/hello"]
```

2. run `g++ main.cc -o hello`
3. run `sudo docker build . -t hello`
4. run `sudo docker run -it -p8080:80 hello:latest`
5. Observe the error.

### Expected behavior

`sudo docker run -it -p8080:80 hello:latest` succeeds.

### docker version

```bash
Client: Docker Engine - Community
 Version:           23.0.2
 API version:       1.42
 Go version:        go1.19.7
 Git commit:        569dd73
 Built:             Mon Mar 27 16:16:40 2023
 OS/Arch:           linux/amd64
 Context:           default

Server: Docker Engine - Community
 Engine:
  Version:          23.0.2
  API version:      1.42 (minimum version 1.12)
  Go version:       go1.19.7
  Git commit:       219f21b
  Built:            Mon Mar 27 16:16:40 2023
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.6.19
  GitCommit:        1e1ea6e986c6c86565bc33d52e34b81b3e2bc71f
 runc:
  Version:          1.1.4
  GitCommit:        v1.1.4-0-g5fd4c4d
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

### docker info

```bash
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc.)
    Version:  v0.10.4
    Path:     /usr/libexec/docker/cli-plugins/docker-buildx
  compose: Docker Compose (Docker Inc.)
    Version:  v2.17.2
    Path:     /usr/libexec/docker/cli-plugins/docker-compose
  scan: Docker Scan (Docker Inc.)
    Version:  v0.23.0
    Path:     /usr/libexec/docker/cli-plugins/docker-scan

Server:
 Containers: 9
  Running: 0
  Paused: 0
  Stopped: 9
 Images: 5
 Server Version: 23.0.2
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Using metacopy: false
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: systemd
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 1e1ea6e986c6c86565bc33d52e34b81b3e2bc71f
 runc version: v1.1.4-0-g5fd4c4d
 init version: de40ad0
 Security Options:
  apparmor
  seccomp
   Profile: builtin
  cgroupns
 Kernel Version: 5.10.0-21-cloud-amd64
 Operating System: Debian GNU/Linux 11 (bullseye)
 OSType: linux
 Architecture: x86_64
 CPUs: 8
 Total Memory: 15.63GiB
 Name: p3achygo-dev-8
 ID: f6e2d016-3703-4911-b8dc-d0a22f42e3c8
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
```

**Comment by thaJeztah** [comment, marked as answer]
Is your binary statically linked? the "not found" may be returned by linux if a linked library cannot be found as well.

**Comment by p3achyjr** [comment]
It worked! Thanks so much :)

What happens if we are in a situation where we cannot produce a fully statically-linked library? Would we have to copy over all library files to the container?

**Comment by cpuguy83** [comment]
@p3achyjr In this case I'm guessing you most likely linked to glibc but alpine use musl-libc.
I'd bet it would work fine if you used a glibc based distro, probably even just busybox.

**Comment by p3achyjr** [comment]
Makes sense--I am also passing `-lstdc++`, so that could contribute as well. Does busybox come prepackaged with C++ static libraries as well?

**Comment by p3achyjr** [comment]
It doesn't seem like busybox comes with libstdc++ by default. By using `gcr.io/distroless/cc-debian11` as a base distro, this worked for me :)
